errors: skip fatal error highlighting on windows

Hakerh400 opened the pull request:
Some consoles do not convert ANSI escape sequences to colors,
rather display them directly to the stdout. On those consoles,
libuv emulates colors by intercepting stdout stream and calling
corresponding Windows API functions for setting console colors.
However, fatal error are handled differently and we cannot easily
highlight them.

Fixes https://github.com/nodejs/node/issues/29387

---

According to [this](https://api.dart.dev/stable/1.24.3/dart-io/Stdout/supportsAnsiEscapes.html), Windows fully supports ANSI escapes starting from v.1607 ("Anniversery Update", OS build 14393), so we could add additional check:

```js
if (process.platform === 'win32') {
  const ver = os.release().split('.').map(a => +a);
  if (ver[0] !== 10 || ver[2] < 14393) {
    return originalStack;
  }
}
```

but given that https://github.com/nodejs/node/issues/29387#issuecomment-620716360 uses build 17763, seems like that this test is not reliable.

---

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

jasnell commented:
Hmm... I understand with and agree with the general nature of this change but it still bothers me. Presumptively, this is an issue for more than just printing the error stack and I really dislike being forced to spread platform specific checks throughout the code as those become harder to maintain (or even keep track of) over time. I'm not sure I have a better suggestion right now but I'm reluctant to +1 this.

BridgeAR commented:
I am also reluctant to completely disable colors on fatal errors in Windows. The issue about the Windows version seems to be around differentiating Windows Server 2019 from regular Windows 10.

If there's any other possibility to detect for example if `cmd.exe` is used or not, we might at least use that in addition.

Hakerh400 commented:
It's interesting how all tests passed, while none of them is modified. I actually see why - this change cannot be detected programmatically on Windows (or at least very hard).

---

>  I really dislike being forced to spread platform specific checks throughout the code as those become harder to maintain (or even keep track of) over time.

@jasnell. I hope we all realize this change is temporary. The actual fix is described in https://github.com/nodejs/node/issues/29387#issuecomment-526918482, but it requires a lot of refactoring and also requires some changes in libuv. Given that this issue has been making troubles for a lot of Windows users for more than a year, wouldn't it be better to disable highlighting while we discuss how to actually resolve the issue?

> I am also reluctant to completely disable colors on fatal errors in Windows.

@BridgeAR. I added additional check, PTAL. In general, there are no official Microsoft documentation regarding ANSI escapes support. There is a [script](https://github.com/keqingrong/supports-ansi/blob/master/index.js) for detecting the ANSI support, but it is not reliable. I tested it on different virtual machines with various console emulators and it does not work in all cases.

Hakerh400 commented:
Any change requests here? If not, can we have this landed before 14.2.0 release (or at least before 14.3.0)?

BridgeAR commented on lib/internal/errors.js:
```diff
@@ -674,6 +674,20 @@ const fatalExceptionStackEnhancers = {
+      if (ver[0] !== 10 || ver[2] < 14393) {
+        return originalStack;
+      }
+    }
```
Instead of adding this here, we could just change the following line: https://github.com/nodejs/node/blob/f446b2058dcfaf9d4e2f6bdfe555bc4437f41acf/lib/internal/tty.js#L150

It would have to be changed to `COLORS_2`.

Hakerh400 replied:
That would unnecessarily break valid `console.log` uses cases too.

BridgeAR replied:
True, I guess in this case it's good as it is.

Hakerh400 commented:
Four tests were failing due to the fact that `error` object does not need to be an instance of `Error` constructor (and may not have `stack` present).

I updated the code and rebased onto master.
All tests seem to pass locally now. Can we have another CI?

Hakerh400 commented:
friendly ping

BridgeAR commented:
Landed in 1c619143eb44 🎉

DerekNonGeneric commented:
Glad this landed! I may be a bit late in asking, but would a `TODO` comment have been helpful to keep track of the fact that this is supposed to be a temporary fix and not a permanent solution?
